exonum/src/explorer/mod.rs

```diff
@@ -466,25 +547,121 @@ impl CommittedTransaction {
 ///
 /// - `type` field contains transaction type (`"in-pool"`).
 /// - `content` is JSON serialization of the transaction.
-#[derive(Debug, Serialize)]
-#[serde(tag = "type", rename_all = "kebab-case")]
-pub enum TransactionInfo {
+///
+/// # Examples
+///
+/// Use of the custom type parameter for deserialization:
+///
+/// ```
+/// # #[macro_use] extern crate exonum;
+/// # #[macro_use] extern crate serde_json;
+/// # use exonum::blockchain::{ExecutionResult, Transaction};
+/// # use exonum::crypto::{PublicKey, Signature};
+/// # use exonum::explorer::TransactionInfo;
+/// # use exonum::storage::Fork;
+/// transactions! {
+///     Transactions {
+///         const SERVICE_ID = 1000;
+///
+///         struct CreateWallet {
+///             public_key: &PublicKey,
+///             name: &str,
+///         }
+///         // other transaction types...
+///     }
+/// }
+/// # impl Transaction for CreateWallet {
+/// #     fn verify(&self) -> bool { true }
+/// #     fn execute(&self, _: &mut Fork) -> ExecutionResult { Ok(()) }
+/// # }
+///
+/// # fn main() {
+/// let json = json!({
+///     "type": "in-pool",
+///     "content": {
+///         "protocol_version": 0,
+///         "service_id": 1000,
+///         "message_id": 0,
+///         "body": {
+///             "public_key": // ...
+/// #                         PublicKey::zero(),
+///             "name": "Alice"
+///         },
+///         "signature": // ...
+/// #                    Signature::zero()
+///     }
+/// });
+///
+/// let parsed: TransactionInfo<CreateWallet> = serde_json::from_value(json).unwrap();
+/// assert!(parsed.is_in_pool());
+/// assert_eq!(parsed.content().name(), "Alice");
+/// # } // main
+/// ```
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(tag = "type", rename_all = "kebab-case", bound(serialize = "T: SerializeContent"))]
+pub enum TransactionInfo<T = Box<Transaction>> {
     /// Transaction is in the memory pool, but not yet committed to the blockchain.
     InPool {
         /// Transaction contents.
-        #[serde(serialize_with = "CommittedTransaction::serialize_content")]
-        content: Box<Transaction>,
+        #[serde(serialize_with = "SerializeContent::serialize_content")]
+        content: T,
     },
 
     /// Transaction is already committed to the blockchain.
-    Committed(CommittedTransaction),
+    Committed(CommittedTransaction<T>),
+}
+
+/// A helper trait functionally equivalent to `serde`'s `Serialize`.
+///
+/// The trait is used to specify bounds on the `Serialize` implementation
+/// in transaction-related types in the `explorer` module, such as [`TransactionInfo`]
+/// and [`CommittedTransaction`].
+///
+/// # Why separate trait?
+///
+/// It is impossible to implement `Serialize` for `Box<Transaction>` (per Rust restrictions).
+/// Similarly, it is impossible to specify `Serialize` as a super-trait for `Transaction`,
```

Comment:
But in theory we can use type erased serde wrapper for serialize.
https://docs.rs/erased-serde/0.3/erased_serde/trait.Serialize.html

Reply:
Yeah, I suppose; but it would require deeper changes (refactoring `Transaction`) than this PR. As serialization refactoring is currently underway, I think that it could make hacks / questionable code like this one obsolete altogether.